### ISSUE TITLE
KLpq biased

### ISSUE THREAD

**Issue by naesseth** [blei-lab/edward#296]
The KL(P||Q) as implemented [here](https://github.com/blei-lab/edward/blob/master/edward/inferences/klpq.py) seems to be biased. It uses the self-normalized importance sampler which is not unbiased

It is possible to get unbiased gradients by noting that minimizing KL(P||Q) is equivalent to maximizing `\int p(x,z) log q(z) dz`

**Comment by dustinvtran** [comment]
Yeah it's a biased (but consistent) estimate of the true gradient. Can you add more details on how to obtain unbiasedness?

**Comment by dustinvtran** [comment]
oh wait i see what you mean. basically you're arguing for unbiased gradients for a different objective. sure, that makes sense. the only algorithmic change is to no longer normalize.

**Comment by naesseth** [comment]
Exactly, this will make sure that even for finite number of samples you are optimizing the KL(P||Q). Optimizing the different objective is equivalent with the original one.

**Comment by naesseth** [comment]
Although I would expect that as the dimension of the latent variable increases we would still have to normalize due to variance problems... The normalization part is basically estimating a constant, p(x) in fact, and is known to significantly reduce variance at the cost of some bias. This means that at the same time we would get a stochastic lower bound on log p(x), i.e. `log N^{-1} \sum_i p(x,z_i)/q(z_i)`, or ofc an unbiased estimate of p(x) if we just skip the logarithm.

**Comment by dustinvtran** [comment]
How about defaulting to the unbiased version, and providing the biased version as an option to the algorithm?

**Comment by naesseth** [comment]
Sounds like a good solution.
